### ISSUE TITLE
encoding bug in OpenJPEG third party module

### ISSUE THREAD

**Issue by zivy** [InsightSoftwareConsortium/ITK#4517]
The version of OpenJPEG in ITK which is used for JPEG2000 compression (default codec when writing DICOM via GDCM) has a bug encoding int32, it silently fails writing (incorrect write). When attempting to read the result the reader will throw an exception reporting the issue (correct behavior). This appears to be resolved in newer versions of OpenJPEG, so likely updating the module to a newer version will resolve the issue. 

Note that the bug was still in OpenJPEG circa 2020 (see [this discussion](https://github.com/zarr-developers/numcodecs/issues/73)).

Python code illustrating the issue:
```
import itk
import numpy as np
from imagecodecs import jpeg2k_encode, jpeg2k_decode

file_name = "slice.dcm"
arr = np.arange(100, dtype="int32").reshape(10,10)

image = itk.GetImageFromArray(arr)
itk.imwrite(image, file_name, compression=True)
itk.imread(file_name)

#encoding/decoding works with the openjpeg version used by the imagecodecs package which relies
#on version 2.5.0 of openjpeg
encoded_arr = jpeg2k_encode(arr)
decoded_arr = jpeg2k_decode(encoded_arr)
```

**Comment by issakomi** [comment]
> The version of OpenJPEG in ITK which is used for JPEG2000 compression (default codec when writing DICOM via GDCM) 

GDCM uses own OpenJPEG library (or _"system"_ optionally), it is `libitkgdcmopenjp2-5.4.a` in ITK build. Yes, it is rather old, 2.3.0 (from 2019, AFAIK). CC @malaterre

**Comment by blowekamp** [comment]
Is this an issue that need to be reported and addressed in the GDCM repo?

**Comment by issakomi** [comment]
> Is this an issue that need to be reported and addressed in the GDCM repo?

> > GDCM uses own OpenJPEG library (or "system" optionally)

Yes, but also there may be other issues with BitsAllocated 32 sometimes and the line
`prec=%u (should be between 1 and 38 according to the JPEG2000 norm. OpenJpeg only supports up to 31)`
is still in OpenJpeg master, s. [here](https://github.com/uclouvain/openjpeg/blob/0fe2a2c591ee9e91dd9ba3334a015386c780e86b/src/lib/openjp2/j2k.c#L2291) and [here](https://github.com/uclouvain/openjpeg/blob/0fe2a2c591ee9e91dd9ba3334a015386c780e86b/src/lib/openjp2/j2k.c#L6621). It is not simple, IMHO. Maybe build GDCM with system's OpenJpeg (on Linux it is easy and the version is 2.5.0+) and test, does it help or not.

**Comment by zivy** [comment]
Just posted the issue on the GDCM bug-tracker (https://sourceforge.net/p/gdcm/bugs/559/). 

@malaterre, really appreciate your help resolving this. We have some naive SimpleITK users who encountered the problem, using a system OpenJPEG is not an option for them - it is beyond what they can do. Thanks.

**Comment by zivy** [comment]
Hello @malaterre, 

Sorry, but I'm a bit confused.

Per @issakomi's [observation](https://github.com/InsightSoftwareConsortium/ITK/issues/4517#issuecomment-1998881962), the OpenJPEG library used by GDCM appears to be from`Modules/ThirdParty/GDCM/src/gdcm/Utilities/gdcmopenjpeg`  which is separate from the `Modules/ThirdParty/OpenJPEG` (I originally thought that module was what needed to be upgraded). This is why I ended up posting the request for updating the library on the gdcm issue tracker. What am I missing here? Thanks for taking a look at this.

**Comment by blowekamp** [comment]
@zivy  Can you test to see if this bug has been addressed with the OpenJPEG updates?

**Comment by hjmjohnson** [comment]
Closing as fixed.

Commit `a6dd845c17` ("ENH: Remove GDCM's OpenJPEG source", 2025-02-07) removed the GDCM-bundled OpenJPEG copy, and ITK now exposes its own up-to-date third-party OpenJPEG to GDCM. The directory `Modules/ThirdParty/GDCM/src/gdcm/Utilities/gdcmopenjpeg/` no longer exists, and `Modules/ThirdParty/OpenJPEG/src/openjpeg/CMakeLists.txt` declares:

```cmake
set(OPENJPEG_VERSION_MAJOR 2)
set(OPENJPEG_VERSION_MINOR 5)
set(OPENJPEG_VERSION_BUILD 4)
```

This was the remediation path the original report (and @blowekamp's January 2025 follow-up) predicted would fix the int32 silent-write-failure encoding bug. If you still see corrupted output on a current ITK build please re-open with a fresh reproducer against the OpenJPEG 2.5.4 path.

Triaged as part of the 2026-04-10 open-issue relevance review.
